sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java

```diff
@@ -788,74 +518,66 @@ private CoderFactory getDefaultCoderFactory(Class<?> clazz) throws CannotProvide
+    return result;
   }
 
   /**
-   * Returns the {@link Coder} to use by default for values of the given
+   * Creates a {@link Coder} to use for values of the given
    * parameterized type, in a context where the given types use the
    * given {@link Coder Coders}.
    *
    * @throws CannotProvideCoderException if no coder can be provided
    */
-  private Coder<?> getDefaultCoder(
+  private Coder<?> createCoderFromParameterizedType(
       ParameterizedType type,
       Map<Type, Coder<?>> typeCoderBindings)
           throws CannotProvideCoderException {
 
-    CannotProvideCoderException factoryException;
-    try {
-      return getDefaultCoderFromFactory(type, typeCoderBindings);
-    } catch (CannotProvideCoderException exc) {
-      factoryException = exc;
+    List<Coder<?>> typeArgumentCoders = new ArrayList<>();
+    for (Type typeArgument : type.getActualTypeArguments()) {
+      try {
+        Coder<?> typeArgumentCoder =
+            createCoderFromTypeDescriptor(TypeDescriptor.of(typeArgument), typeCoderBindings);
+        typeArgumentCoders.add(typeArgumentCoder);
+      } catch (CannotProvideCoderException exc) {
+        throw new CannotProvideCoderException(
+            String.format("Cannot provide coder for parameterized type %s: %s",
+                type,
+                exc.getMessage()),
+            exc);
+      }
     }
+    return createCoderFromFactories(TypeDescriptor.of(type), typeArgumentCoders);
+  }
 
-    CannotProvideCoderException annotationException;
-    try {
-      Class<?> rawClazz = (Class<?>) type.getRawType();
-      return getDefaultCoderFromAnnotation(rawClazz);
-    } catch (CannotProvideCoderException exc) {
-      annotationException = exc;
+  /**
+   * Attempts to create a {@link Coder} from any registered {@link CoderFactory} returning
+   * the first successfully created instance.
+   */
+  private Coder<?> createCoderFromFactories(
+      TypeDescriptor typeDescriptor, List<Coder<?>> typeArgumentCoders)
+      throws CannotProvideCoderException {
+    for (CoderFactory coderFactory : coderFactories) {
+      try {
+        return coderFactory.create(typeDescriptor, typeArgumentCoders);
+      } catch (CannotProvideCoderException e) {
+        // This coder factory is unable to provide a coder, move onto the next one.
+      }
     }
 
     // Build up the error message and list of causes.
     StringBuilder messageBuilder = new StringBuilder()
-        .append("Unable to provide a default Coder for ").append(type)
-        .append(". Correct one of the following root causes:");
-
-    messageBuilder
-        .append("\n  Building a Coder using a registered CoderFactory failed: ")
-        .append(factoryException.getMessage());
-
-    messageBuilder
-        .append("\n  Building a Coder from the @DefaultCoder annotation failed: ")
-        .append(annotationException.getMessage());
+        .append("Unable to provide a default Coder for ").append(typeDescriptor)
```

Comment:
I guess this error message is a bit long now that there are not multiple causes possible. The registered factories should be listed at least, though it is definitely too much to list all of the reasons. That's a major loss of usability, though I don't see a way around it.